Add NoSQL topic

ang-zeyu opened the pull request:
Closes #130

ang-zeyu committed:
Add NoSQL topic

nbriannl commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,370 @@
+In contrast to tables and table entries in SQL databases, we typically have collections and documents.
+A database comprises of multiple collections, which in turn consists of multiple documents.
+
+In a simplistic e-commerce website for example, you may have the following collections:
```
In a ~~simplistic~~ **simplified** e-commerce website for example, you may have the following collections:

nbriannl commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,370 @@
+
+In a simplistic e-commerce website for example, you may have the following collections:
+- customers - storing the account details of customers, their purchase histories, etc.
+- items - a collection of all items (documents) available for purchase
```
items - a collection of all items (**which are** documents) available for purchase

nbriannl commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,370 @@
+- admin - a collection storing admin account details
+- **...**
+
+The following might be how an item available for purchase under the *items* collection
```
The following might be how an item **document** available for purchase under the *items* collection

nbriannl commented on contents/data/nosql/NoSQL.md:
````diff
@@ -0,0 +1,370 @@
+A potential new user sends a POST request to register a new account, and your server would then simply call
+the following mongoDB code to store the details after some processing.
+
+```js
+db.customers.insertOne({
+  username: "panda",
+  password: hashedPassword,
+  email: "panda@pandas.com"
+})
+```
````
As someone who is only at basic familiarity with MongoDB or NoSQL in general, the object we intend to put in can have just part of the entire schema filled in? Just a clarification

nbriannl replied:
An example of how the customers collection is structured would be useful if i think about it.

ang-zeyu replied:
Hmm, I mentioned this in the **schemas in NoSQL** section later, but yes including some preface of how the customers collection is structured would be nice 👍

nbriannl commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,371 @@
+
+Here, the code simply provides a query JSON object matching the structure of the items collection.
+Specifically, items with a price of less than 30 are filtered with the `$lt: 30` key-value pair,
+and items with a popularity of more than 8 are filtered with the `$gt: 8`.
```
Is the above code block the json object in response to a query? If yes, It would help if there was a sample of the query as well.

ang-zeyu replied:
I discarded an entire chunk above by accident 😅

nbriannl commented on contents/data/nosql/NoSQL.md:
````diff
@@ -0,0 +1,371 @@
+  popularity: 9.7,
+  _id: 9d1793bd491349n913847n93d
+}
+```
````
Personal opinion, but introducing the this code block, then showing the code block for cart items is more intuitive for me.

```js
...
cartItems: [
  "9d1793bd491349n913847n93d",
  "9d1793bd491349n9155321de3",
]
...
```

I had to put two and two together and realize that "probably itemId1 and itemId2 are supposed to be _id later on."

ang-zeyu replied:
makes sense.

nbriannl reviewed (commented):
The latter sections are very clear and reads well. "How to get started with NoSQL?" was good. Just some comments. I only used mongo for my orbital so I'm reading this guide from a novice's perspective.

ang-zeyu committed:
Address review comments

ang-zeyu commented:
> The latter sections are very clear and reads well. "How to get started with NoSQL?" was good. Just some comments. I only used mongo for my orbital so I'm reading this guide from a novice's perspective.

Thanks for the review @nbriannl! I've addressed the relavant parts

jamessspanggg commented on _markbind/navigation/mainNav.md:
```diff
@@ -79,6 +79,7 @@
   * Databases
     * [DBMS]({{baseUrl}}/contents/data/databases/databases.html)
     * [SQL]({{baseUrl}}/contents/data/sql/SQL.html)
+	* [NoSQL]({{baseUrl}}/contents/data/nosql/NoSQL.html)
```
The indentation can be at the same level as SQL

ang-zeyu replied:
used unexpanded tabs instead of spaces here 🤦‍♂

jamessspanggg commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,400 @@
+
+This means that each individual document has no restriction on what keys it must have,
+the number of keys, the type of values, etc.
+(Note the missing fields for the second item in the earlier code example [above](#schemaless), which are intentionally ommitted.)
```
ommitted -> omitted

jamessspanggg commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,400 @@
+  cart: {
+    totalPrice: 100,
+    cartItems: [ ... ],
+	discountCode: "panda"
```
is the indentation on purpose?

jamessspanggg commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,400 @@
+Such simple relations are extremely suited to and expressable with the schema-less characteristic
+of NoSQL databases.
+
+In the case of more complicated many-to-many relationships, references are commonly stored using references,
```
do you mean `relations are commonly stored using references`?

jamessspanggg commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,400 @@
+As explained above, NoSQL databases are easily horizontally scalable.
+
+As businesses grow and see increasing amounts of web traffic, it is crucial that its databases can scale to meet
+consumer and buisness demands.
```
business*

jamessspanggg commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,400 @@
+
+Congratulations! You should have a good handle on mongoDB by now, and learnt how you would generally interact with NoSQL databases.
+
+If you're interested in learning more about mongoDB, I recommend going through the following topics in order in the above resources:
```
I think `...going through the following topics in order.` would be sufficient

jamessspanggg reviewed (commented):
The chapter was very clear from a beginner's standpoint, just some nits.

ang-zeyu committed:
Add some spelling and grammar fixes

ang-zeyu committed:
Address review comments

ang-zeyu commented:
> The chapter was very clear from a beginner's standpoint, just some nits.

Thanks for the review as well @jamessspanggg! 😆 
I've addressed the relavant parts

nbriannl reviewed (approved):
LGTM
![LGTM](https://media.giphy.com/media/3o7abB06u9bNzA8lu8/giphy.gif)

ang-zeyu committed:
Minor grammar fix

ang-zeyu committed:
Add caption for survey chart

openorclose commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,407 @@
+  type: "book",
+  price: 20,
+  popularity: 9.7,
+  _id: 9d1793bd491349n913847n93d
```
Hm I'm also new to nosql, but over here _id is some kind of unquoted string, but later on it's inside quotes ""s.

Are they interchangeable? I was a bit confused at this part.

ang-zeyu replied:
Good catch, thanks!

ang-zeyu committed:
Address review comment

ang-zeyu committed:
Fix more unexpanded tabs

ang-zeyu committed:
Update reviewers

ang-zeyu committed:
Add table of contents

yash-chowdhary commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+Specifically, items with a price of less than 30 are filtered with the `$lt: 30` key-value pair,
+and items with a popularity of more than 8 are filtered with the `$gt: 8`.<br>
+Subsequently, a "cursor" is returned, which undergoes a `sort` operation by the respective fields of `price`
```
I'm just wondering if the "cursor" could be explained a bit. A newcomer could be thrown off by the lingo. 
Perhaps a short explanation in a tooltip/modal could help as the user wouldn't need to leave the page to google it. Just a suggestion, though ✌️

ang-zeyu replied:
Thanks for the review @yash-chowdhary!

Good point, its actually just your standard iterator through a collection, though mongoDB prefers to call it cursor. Will update this.

yash-chowdhary commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+
+
+<box type="info" icon-size="2x">
```
Just wondering if we should standardize the box themes in the page? Perhaps we can stick to `light`? 
I only bring it up because there are occurrences of the same type of box with different themes.

ang-zeyu replied:
Hmm the different designs are intentional; For example the last box in particular with `light` makes the section look a little wordy, using the non `light` version makes it seem less so since it contrasts with the surrounding better.

That said its subjective; I'm open to standardising it later on if requested 😄

yash-chowdhary reviewed (commented):
Looks great! Minor comments.

ang-zeyu committed:
Address review comments

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,436 @@
+
+Some common implementations of such databases are key-value stores, document stores (which are an extension of key-value stores),
+and though less common - graph databases (which can also be considered an extension of document stores).
+
```
If the reader can understand this section, most likely she is knowledgeable enough in the database domain to know what NoSQL is. Tech terms density is very high too. Need to make the introduction gentler and gradual. You can also declare your assumptions about prior knowledge at the start.
One option is to declare the SQL chapter is a prerequisite, and write this chapter as if the reader is a beginner but has read the SQL chapter.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,436 @@
+CRUD operations with NoSQL databases are just as powerful as traditional SQL databases, and in some
+cases even more succinct.
+
+<br>
```
What's the point you are trying to make in this section? If the point is the following, perhaps the description above need not go into so much low level details?
>CRUD operations with NoSQL databases are just as powerful as traditional SQL databases, and in some
cases even more succinct.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,436 @@
+
+---
+
+## How does NoSQL work?
```
Perhaps you can name this section `Key characteristics`? Then, give 2-3 sub sections:
* Key characteristic 1: ...
* Key characteristic 2: ...

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,436 @@
+
+<br>
+
+### Horizontal scaling :fas-server:
```
Move this to the WHY section? Best not to repeat the same thing twice in one chapter.

damithc reviewed (changes requested):
Good start. The high-level structure needs to be more focused. I suggest the following.

# What is NoSQL

Give an 'outsider' introduction.  Can give a very simple helloworld type example to show the nature of the beast.

# Key characteristics

What makes NoSQL stand out? What makes it different from the alternatives?

##  Key characteristic 1: ...
##  Key characteristic 2: ...
##  Key characteristic 3: ...

# Why NoSQL (or) Pros

## Benefit 1: ...
## Benefit 2: ...

# Why not NoSQL (or) Cons

ang-zeyu committed:
Address review comments

ang-zeyu commented:
@damithc I addressed the comments, and made some extra changes based on them which I think were relavant as well

1)Restructured the article
- What is NoSQL gives a "hello world" preview of a document, which is followed by "how does nosql work - basic structure of document databases", which should give a good preface for the rest of the article.
- The other parts of "how does nosql work" were rewritten as "key characteristics"
- De duplicated the section on horizontal scaling

2)Trimmed some of the other heavier illustrations to give a more "high-level" overview, besides the commented section here

> perhaps the description above need not go into so much low level details?

3)Many more tooltips / popovers for tech terms (eg. `JSON`, `one-to-many`, `many-to-many` ) to remove as much technical barrier to the article as possible
* There's a popover at the very bottom that slips out of view of the page currently, but the positioning should be fixed once https://github.com/MarkBind/markbind/pull/1033 is merged


4)Made some miscellaneous style changes

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+## What is NoSQL?
+
+**Non-Structured Query Language** (**NoSQL**) is a wide set of implementations of query technologies used to retrieve and store data in a non-tabular format.
```
Are you sure NoSQL means Non-Structured Query Language? I've heard other definitions including 'Not only SQL'

ang-zeyu replied:
There's quite a few interpretations of the term, I chose the one that makes the most sense in this context ( given the reader assumption of `sql` )

damithc commented on contents/data/nosql/NoSQL.md:
````diff
@@ -0,0 +1,430 @@
+}
+```
+
+To start, let's jump into the most common implementation of such databases, _document databases_.
````
This implies there are other implementations. Perhaps we should mention a few here? It's comforting to the reader to know what other doors exist before going through one of them.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+---
+
+## How does NoSQL work - basic structure of document based databases
```
No need for this heading as it disrupts the flow unnecessarily. The content fits well into the previous heading i.e., you are giving an example of a NoSQL database to explain the WHAT aspect.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+![](documentDatabaseStructure.png)
+
+In contrast to tables and table entries in SQL databases, we typically have collections and documents.
```
who is 'we'?

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+In a simplified e-commerce website for example, you may have the following collections: { id="main-example" }
+- _customers_ - storing the account details of customers, their purchase histories, etc.
+- _items_ - a collection of all items available for purchase ( which are documents ) 
```
>( which are documents )

No need to pad using spaces?

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+<box type="info" light>
+
+Note the missing fields for the second item in the <em>items</em> collection <a href="#main-example">earlier</a>, which are intentionally omitted.
```
A popover may be better so that the user doesn't have to jump to another location of the document. Use an `include` instead of copy pasting the content though.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+Documents can even contain other documents, arrays, and depending on the implementation, likely anything the database can serialize and deserialize.
+
+#### Validation
```
Let's avoid sudden sub-headings unless they are used consistently in all sister sections.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+### 3. Straightforward expression of relations in data
+
+The world is full of relations. For example, a patient _is related to_ his / her disease record,
```
his/her? i.e., no spaces around '/'. Can also use 'her' or 'their' in place of 'his/her'

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+<br>
+
+### 1. Highly suited for agile development :dash:
```
Try to avoid the word 'agile' as some people don't believe in 'agile' (as in 'agile methods').

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+in NoSQL, this means that developers can adapt the database quicker to changing customer and business requirements.
+
+
+### 2. Horizontal scaling :fas-server:
```
Seems like :dash: shows up in the page nav but not :fas-server: Perhaps not use icons at all?

damithc replied:
Phrase as a benefit? e.g., 'Easy horizontal scaling'

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,430 @@
+
+<div>
+Vertical scaling ( increasing the processing power of the machine ) can only go so far until the single machine hits its limit.</div>
+</box>
```
Use the same style of boxes throughout the entire chapter?

damithc commented on contents/data/nosql/NoSQL.md:
This section seems too long. I seriously doubt someone new to this topic will have the patience to read all this. They'll probably exit this chapter through the first link in this section. See if you can shorten it to about half the current length.

damithc reviewed (changes requested):
Looks good now. Some suggestions added.

ang-zeyu committed:
Address review comments

ang-zeyu commented:
I've addressed the comments @damithc 

The `How to get started section` is still somewhat long, but it feels considerably less wordy now.
I have a few more ideas to trim it still though, if needed.

Tejas2805 reviewed (changes requested):
Can you make the headings follow [titleCase](https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case) since we are aiming for that across the book?

ang-zeyu committed:
Change headings to title case

ang-zeyu commented:
> Can you make the headings follow [titleCase](https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case) since we are aiming for that across the book?

👍 Updated

ang-zeyu committed:
update toc to title case as well

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+
+## What is NoSQL?
+
+**Non-Structured Query Language** (**NoSQL**) is a wide set of implementations of query technologies used to retrieve and store data in a non-tabular format.
```
It's good if you can also mention other names such as 'not only SQL', at least as a tooltip.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+<li>Some of these types can even be used together!</li>
+</box>
+
+One such commonly used database is MongoDB, which is a document based database.
```
This follows the previous sentence better?
>One such commonly used document database is MongoDB.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+One such commonly used database is MongoDB, which is a document based database.
+
+
+![](documentDatabaseStructure.png)
```
Resize the image so that the text size matches the main body text size.

damithc replied:
Also, see if you can redraw the diagram to reduce empty space taken up. e.g., there is a lot of empty space above and below the work `Collections`.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+
+![](documentDatabaseStructure.png)
+
+In contrast to tables and table entries in SQL databases, document dtabases comprise of collections and documents.
```
typo: dtabases

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+![](documentDatabaseStructure.png)
+
+In contrast to tables and table entries in SQL databases, document dtabases comprise of collections and documents.
+A database comprises of multiple collections, which in turn consists of multiple documents.
```
Combine these two sentences? They seem to have some overlap.

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+and JSON-like queries.
+
+Let's get back to the above [example](#main-example) of an e-commerce website. To register a new user account,
+an example query made in mongoDB would be like so:
```
The word _query_ feels like a read operation. Are write operations also called queries?

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+<box type="info" light>
+Note the missing fields for the second item in the <em>items</em> collection
+<popover header="Example items collection of an e-commerce website">earlier,<template slot="content"><include src="NoSQL.md#schemaless-example" /></template></popover>
+which are intentionally omitted.
```
How come the text in this popover appear smaller than regular text? 🤔

ang-zeyu replied:
Changed this to a modal; Popovers are hardcoded to be `< 276px` in bootstrap apparently.  ( for mobile-friendliness )

damithc replied:
Actually, I was referring to the font size. Popovers use smaller font size?

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+For example, in the _customers_ collection, where the fields of a customer are unlikely to
+change, it can be especially helpful to enforce a strict schema on documents; 
+This would prevent the unsuspecting programmer
+from say, deleting a customer's password, which would be disastrous.
```
Well, we don't know if it would be disastrous, but undesirable for sure :-)

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+
+One of the most consequential impacts NoSQL has had was enabling faster iterated development.
+Given the highly flexible relational structure of NoSQL databases, and the schemaless format of documents 
+in NoSQL, this means that developers can adapt the database quicker to changing customer and business requirements.
```
Should we mention the fact that in contrast SQL database are harder to change later due to the difficulty in modifying the schema?

damithc commented on contents/data/nosql/NoSQL.md:
```diff
@@ -0,0 +1,437 @@
+
+<box type="success" light>
+
+As businesses grow and see increasing amounts of web traffic, it is <em>crucial</em> that its databases can scale to meet
```
I wonder why _web traffic_ was specifically mentioned here. It doesn't necessarily have to be web traffic, right?

damithc commented on contents/data/nosql/NoSQL.md:
This is not attached to any sentence:
>(Amazon uses a proprietary NoSQL database!)

damithc commented on contents/data/nosql/NoSQL.md:
users might not know which ones are NoSQL. Also, this chart does not show any trends over time.

damithc commented on contents/data/nosql/NoSQL.md:
_flavors_ refer to what?

damithc commented on contents/data/nosql/NoSQL.md:
many-to-many?

damithc commented on contents/data/nosql/NoSQL.md:
try your hand?

damithc commented on contents/data/nosql/NoSQL.md:
This phrasing implies that the reader has come back to the article after learning MongoDB, which is unlikely.

damithc reviewed (changes requested):
The structure looks good. Some cosmetic suggestions added.

ang-zeyu committed:
Address review comments

ang-zeyu committed:
Adjust horizontal scaling diagram and example

ang-zeyu commented:
Ready for another look @damithc 

I've made the suggested changes, and also tweaked the _horizontal scaling_ code example and diagram a little more

damithc reviewed (commented):
Looks mostly ready now. Two things you can consider:
* In some places, you compare NoSQL with SQL. Should that be NoSQL vs Relational? SQL is a language.
* When explaining something new, you can consider starting with something the reader knows, and then guiding the reader to a new place. e.g.,
  >In relational databases, it's hard to do X. But because NoSQL databases are like this, X becomes easier.
  >[instead of] In NoSQL, X is easier. That's because they are like this. In contrast, X is harder in relational databases.
  
   It's a style choice. I'll let you decide.

ang-zeyu committed:
change SQL references to relational databases

ang-zeyu commented:
I've updated the references to `SQL` to `relational` where appropriate.

> When explaining something new, you can consider starting with something the reader knows, and then guiding the reader to a new place. e.g.,

Hmm, I tried changing one of the sections and it feels a little strange

> Relational databases mainly use SQL, having a syntax that is mostly standardised across its different implementations.
> This is in stark contrast to NoSQL databases which can vary from one solution to another greatly. This may incur extra development costs in projects when there is a need to migrate to another solution, or when new developers are introduced to the project.

This ( personally ) shifts the focus of the section ( and the article ) to SQL instead of NoSQL. I think the topic sentence should remain tied to NoSQL so that the focus of the article isn't lost

damithc commented:
> This ( personally ) shifts the focus of the section ( and the article ) to SQL instead of NoSQL. I think the topic sentence should remain tied to NoSQL so that the focus of the article isn't lost

👍

damithc commented:
@ang-zeyu is this ready to be merged?

ang-zeyu committed:
Fix spelling nit

ang-zeyu committed:
Merge branch 'master' into nosql

ang-zeyu commented:
> @ang-zeyu is this ready to be merged?

Fixed one more spelling nit, ready to be merged now!
